versioned/persist/adapter/src/main/java/org/projectnessie/versioned/persist/adapter/spi/AbstractDatabaseAdapter.java

```diff
@@ -825,6 +827,120 @@ public final CommitLogEntry fetchFromCommitLog(OP_CONTEXT ctx, Hash hash) {
 
   protected abstract CommitLogEntry doFetchFromCommitLog(OP_CONTEXT ctx, Hash hash);
 
+  @Override
+  public Stream<CommitLogEntry> scanAllCommitLogEntries() {
```

Comment:
WDYT about (optionally) providing a chronologically ordered stream of commits? That would reduce the amount of in-memory state, I guess (pruning non-heads sooner). Storage would have to maintain an index, which may not always be feasible, but perhaps in some cases?

Reply:
Doubt it could work in all cases - like these:
* (Wall) clock drift could still let children appear before parents (parent create-timestamp < child create-timestamp)
* Indexes in non-tx databases are rather "hopefully consistent"

It's a tricky issue here and I'm not sure there's a way to get around it. Making that `Set<Hash> parents` a bloom filter is also not really going to work, because it will result in false-positives: non-head commits appearing as heads in the function's result. Which in turn would result in false-positives from `identifyReferencedAndUnreferencedHeads()`: returning those as unreferenced heads, which is risky, beacuse Nessie GC could use different expiry mechanisms, expire content quicker, leading to unavailable user data.

The "heap math" for the `Set<Hash> parents` is this:
* It's an open addressing hash set, with a load-factor of .65.
* Each `Hash` (of a 256 bit hash) has an on-heap size of 48 bytes: (up to) 16 bytes Java object header + 32 bytes for the 256 bits.
* The `parents` set size for e.g. 650000 objects would be an `Object[1048576]` (~ 8M heap) plus the `Hash`s (~32M) - so it's not too bad.
* The remaining objects are either short lived or small-ish collections.

Maybe we can go with this approach and see how big repositories will actually become?

Reply:
Using an open addressing hash set SGTM :+1:

Reply:
In theory, we could implement our custom open-addressing-set for 256-bit hashes, which would, based on the load-factor of .65, roughly reduce the heap pressure by the same amount that the `Object[]` occupies - and also have a better memory access pattern. It's an option - but likely none for yet.